packages/mini-oxygen/src/vite/deps-optimizer.ts

```diff
@@ -0,0 +1,113 @@
+import type {ServerResponse} from 'node:http';
+import {readFile} from 'node:fs/promises';
+import type {ViteDevServer} from 'vite';
+import {getErrorPage} from '../common/error-page.js';
+import type {Response} from '../worker/index.js';
+
+export function isEntrypointError(webResponse: Response) {
+  return (
+    webResponse.status === 503 &&
+    webResponse.statusText === 'executeEntrypoint error'
+  );
+}
+
+export type CustomEntryPointErrorHandler = (params: {
+  optimizableDependency?: string;
+  stack: string;
+}) => void | Promise<{
+  status?: number;
+  body?: string;
+  headers?: Record<string, string>;
+} | void>;
+
+export async function handleEntrypointError(
+  viteDevServer: ViteDevServer,
+  webResponse: Response,
+  res: ServerResponse,
+  entryPointErrorHandler?: CustomEntryPointErrorHandler,
+) {
+  const stack = (await webResponse.text())
+    .split('\n')
+    .filter((line) => !line.includes('mini-oxygen'))
+    .join('\n');
+
+  const optimizableDependency = await findOptimizableDependency(
+    viteDevServer,
+    stack,
+  );
+
+  const header = `MiniOxygen could not load the app's entry point.`;
+  const message = optimizableDependency
+    ? `Try adding <code>${optimizableDependency}</code> to the <code>ssr.optimizeDeps.include</code> array in your Vite config.`
+    : '';
+
+  if (!entryPointErrorHandler) {
+    console.warn(
+      '\nWarning: ' + header + '\n' + message.replace(/<\/?code>/g, '"'),
+    );
+  }
+
+  const result = await entryPointErrorHandler?.({optimizableDependency, stack});
```

Comment:
This `entryPointErrorHandler` is passed from the CLI to the mini-oxygen plugin, and it's what writes to `vite.config.js` (it has access to prettier and ast-grep).